### PR TITLE
Integrated contracts-v2

### PR DESCRIPTION
**Pull request by smartcontracts**
## Description
Lots of changes, but the primary ones are:
* Changing how we parse transactions to convert them to "ovm messages"
* Updating `state_manager.go` with the new API
* Updating our logging statements
* Updating our state dump (I'd rather we pull this from somewhere external)

## Contributing Agreement

- [x] I have read and understood the [Optimism Contributing Guide and Code of Conduct](./CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

File: core/state_transition.go [hunk 1]
```diff
@@ -149,6 +138,18 @@ func NewStateTransition(evm *vm.EVM, msg Message, gp *GasPool) *StateTransition
...
 // state and would never be accepted within a block.
 func ApplyMessage(evm *vm.EVM, msg Message, gp *GasPool) ([]byte, uint64, bool, error) {
+	if vm.UsingOVM {
+		// OVM_ENABLED
+		ovmmsg, err := toExecutionManagerRun(evm, msg)
+		if err != nil {
+			return nil, 0, false, err
+		}
+
+		evm.Context.Origin = msg.From()
```
Comment (smartcontracts) [inline core/state_transition.go]: Hopefully this will be made unnecessary by a change @K-Ho is making to the `OVM_ExecutionManager`. Basically needed so I can get `msg.sender == ovmStateManager.owner()` to return true.

File: core/state_transition.go [hunk 2]
```diff
@@ -201,9 +202,16 @@ func (st *StateTransition) preCheck() error {
...
-	if err = st.preCheck(); err != nil {
-		return
+	if !vm.UsingOVM {
+		// OVM_DISABLED
+		if err = st.preCheck(); err != nil {
+			return
+		}
+	} else {
+		// OVM_ENABLED
+		st.buyGas()
```
Comment (smartcontracts) [inline core/state_transition.go]: This is the only part of the `preCheck` that we actually care about.

File: core/state_transition.go [hunk 3]
```diff
@@ -220,80 +228,35 @@ func (st *StateTransition) TransitionDb() (ret []byte, usedGas uint64, failed bo
...
-	l1MessageSender := msg.L1MessageSender()
-	if l1MessageSender == nil {
-		addr := common.HexToAddress("")
-		l1MessageSender = &addr
-	}
-
-	log.Debug("Applying transaction", "from", sender.Address().Hex(), "to", to, "nonce", msg.Nonce(), "l1MessageSender", l1MessageSender.Hex(), "data", hexutil.Encode(msg.Data()))
-
 	if contractCreation {
-		// Here we are going to call the EM directly
```
Comment (smartcontracts) [inline core/state_transition.go:252]: All of this is now handled in `toExecutionManagerRun`, so not needed here anymore.

File: core/state_transition.go [hunk 4]
```diff
@@ -220,80 +228,35 @@ func (st *StateTransition) TransitionDb() (ret []byte, usedGas uint64, failed bo
...
+			st.state.SetNonce(msg.From(), st.state.GetNonce(msg.From())+1)
+		}
+		ret, st.gas, vmerr = st.evm.Call(sender, st.to(), st.data, st.gas, st.value)
 	}
-	if vmerr != nil {
-		log.Debug("VM returned with error", "err", vmerr)
-
-		// If the tx fails we won't have incremented the nonce. In this case, increment it manually
-		log.Debug("Incrementing nonce due to transaction failure")
-		st.state.SetNonce(msg.From(), st.state.GetNonce(sender.Address())+1)
```
Comment (smartcontracts) [inline core/state_transition.go:284]: We need to figure out the best way to handle this nonce eviction thing. @karlfloersch should we just go ahead and update the eviction logic?

File: core/state_transition.go [hunk 5]
```diff
@@ -220,80 +228,35 @@ func (st *StateTransition) TransitionDb() (ret []byte, usedGas uint64, failed bo
...
 		}
 	}
+
 	st.refundGas()
-	st.state.AddBalance(st.evm.Coinbase, new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), st.gasPrice))
 
-	log.Debug("return data", "data", hexutil.Encode(ret))
+	if !vm.UsingOVM {
+		// OVM_DISABLED
+		st.state.AddBalance(st.evm.Coinbase, new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), st.gasPrice))
```
Comment (smartcontracts) [inline core/state_transition.go:266]: No balances in the OVM.

File: core/state_transition_ovm.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+	"bytes"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+)
+
+var GodAddress = common.HexToAddress("0x444400000000000000000000000000000000000")
```
Comment (smartcontracts) [inline core/state_transition_ovm.go]: This is a temporary value -- was a bit confused about our actual plans for this address.

File: core/state_transition_ovm.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+	L1TxOrigin    common.Address "json:\"l1TxOrigin\""
+	Entrypoint    common.Address "json:\"entrypoint\""
+	GasLimit      *big.Int       "json:\"gasLimit\""
+	Data          []uint8        "json:\"data\""
+}
+
+func toExecutionManagerRun(evm *vm.EVM, msg Message) (Message, error) {
+	tx := ovmTransaction{
+		evm.Context.Time,
+		evm.Context.BlockNumber, // TODO (what's the correct block number?)
```
Comment (smartcontracts) [inline core/state_transition_ovm.go:46]: @karlfloersch @tynes what's the correct value for this?

File: core/state_transition_ovm.go [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+	// Signature type
+	data.WriteByte(byte(sigtype)) // 1 byte: 00 == EOACreate, 01 == EIP 155, 02 == ETH Sign Message
+
+	// Signature data
+	data.Write(v.FillBytes(make([]byte, 1, 1)))   // 1 byte: Signature `v` parameter
+	data.Write(r.FillBytes(make([]byte, 32, 32))) // 32 bytes: Signature `r` parameter
+	data.Write(s.FillBytes(make([]byte, 32, 32))) // 32 bytes: Signature `s` parameter
+
+	if sigtype == 0 {
```
Comment (smartcontracts) [inline core/state_transition_ovm.go]: I guess we can get rid of this now, right?

File: core/state_transition_ovm.go [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+	)
+
+	if err != nil {
+		return msg, err
+	}
+
+	return outmsg, nil
+}
+
+func EncodeFakeMessage(
```
Comment (smartcontracts) [inline core/state_transition_ovm.go:141]: We need a better way to do this, will talk with @karlfloersch and @K-Ho

File: core/state_transition_ovm.go [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+		msg.L1MessageSender(),
+		msg.L1RollupTxId(),
+		queueOrigin,
+		msg.SignatureHashType(),
+	)
+
+	return outmsg, nil
+}
+
+func getSignatureType(
```
Comment (smartcontracts) [inline core/state_transition_ovm.go:196]: Only necessary because our sig hash types don't match up with the values in the decompressor contract (or maybe they do now?)

File: core/types/transaction.go [hunk 11]
```diff
@@ -326,6 +326,13 @@ func (tx *Transaction) AsMessage(s Signer) (Message, error) {
 
 	var err error
 	msg.from, err = Sender(s, tx)
+
+	if tx.meta.L1MessageSender != nil {
```
Comment (smartcontracts) [inline core/types/transaction.go:335]: Is this dangerous?

File: core/vm/evm.go [hunk 12]
```diff
@@ -47,14 +47,33 @@ type (
...
+				}
+			}
+		}
+
+		if isUnknown {
+			log.Debug("Calling Unknown Contract", "Address", contract.Address().Hex())
+		}
+
+		if contract.Address() == OvmStateDump.Accounts["OVM_SafetyChecker"].Address {
+			return common.FromHex("0x0000000000000000000000000000000000000000000000000000000000000001"), nil
```
Comment (smartcontracts) [inline core/vm/evm.go]: Temporary: have the safety checker always return true.

File: core/vm/evm.go [hunk 13]
```diff
@@ -202,7 +227,23 @@ func (evm *EVM) Interpreter() Interpreter {
...
+	var isTarget = false
+	if UsingOVM {
+		// OVM_ENABLED
+		if evm.depth == 0 {
+			evm.OriginalTargetAddress = nil
+			evm.OriginalTargetResult = []byte("00")
+		}
+
+		if caller.Address() == OvmExecutionManager.Address &&
+			!strings.HasPrefix(strings.ToLower(addr.Hex()), "0xdeaddeaddeaddeaddeaddeaddeaddeaddead") &&
```
Comment (smartcontracts) [inline core/vm/evm.go:250]: `0xdead` and `0x0000` addresses can't be the target.

File: core/vm/evm.go [hunk 14]
```diff
@@ -211,15 +252,20 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
...
 		return nil, gas, ErrDepth
 	}
-	// Fail if we're trying to transfer more than the available balance
-	if !evm.Context.CanTransfer(evm.StateDB, caller.Address(), value) {
-		return nil, gas, ErrInsufficientBalance
+
+	if !UsingOVM {
+		// OVM_DISABLED
+		// Fail if we're trying to transfer more than the available balance
+		if !evm.Context.CanTransfer(evm.StateDB, caller.Address(), value) {
```
Comment (smartcontracts) [inline core/vm/evm.go:274]: Breaks things in the OVM (don't need it anyway).

File: core/vm/evm.go [hunk 15]
```diff
@@ -266,6 +317,19 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
 			contract.UseGas(contract.Gas)
 		}
 	}
+
+	if UsingOVM {
+		// OVM_ENABLED
+		if isTarget {
+			evm.OriginalTargetResult = ret[96:]
```
Comment (smartcontracts) [inline core/vm/evm.go]: Ignore the first 96 bytes (the return status and ABI encoding parameters).

File: core/vm/interpreter.go [hunk 16]
```diff
@@ -124,6 +124,8 @@ func NewEVMInterpreter(evm *EVM, cfg Config) *EVMInterpreter {
 	}
 }
 
+var returnDataCopy string
```
Comment (smartcontracts) [inline core/vm/interpreter.go:127]: This is the hack that bypasses an apparent bug

File: internal/ethapi/api.go [hunk 17]
```diff
@@ -806,7 +806,7 @@ func DoCall(ctx context.Context, b Backend, args CallArgs, blockNrOrHash rpc.Blo
 		}
 	}
 	// Set default gas & gas price if none were set
-	gas := uint64(math.MaxUint64 / 2)
+	gas := uint64(10000000)
```
Comment (smartcontracts) [inline internal/ethapi/api.go:809]: Will break with max gas because EM asserts that gas is < `maxGasPerQueuePerEpoch` (not sure what the right value is to put here -- block gas limit?)

File: core/state_transition_ovm.go (hunk 6 above)
Comment (tynes) [reply core/state_transition_ovm.go]: The god address is pulled in through this CLI flag - https://github.com/ethereum-optimism/go-ethereum/blob/b2e6790696a97e5e7c7c0c7f6facd4b66478d51f/cmd/utils/flags.go#L789

The plan is to remove it, but then we need a code path in the transaction ingestion pipeline that skips signature verification, since L1 did the signature verification. We talked about not incrementing any nonces for these kinds of txs. Would it make sense to get the L1 tx and compute the `from` and increment that nonce?

File: core/state_transition_ovm.go [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+)
+
+var GodAddress = common.HexToAddress("0x444400000000000000000000000000000000000")
+var ZeroAddress = common.HexToAddress("0x0000000000000000000000000000000000000000")
+
+type ovmTransaction struct {
+	Timestamp     *big.Int       "json:\"timestamp\""
```
Comment (tynes) [inline core/state_transition_ovm.go:19]: Why the escaped `"`?

File: core/state_transition_ovm.go (hunk 7 above)
Comment (tynes) [reply core/state_transition_ovm.go:46]: For an L1 to L2 transaction, is `block.number` supposed to be the L1 block number?

File: core/state_transition_ovm.go [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+
+	var target common.Address
+	if tx.To() == nil {
+		target = ZeroAddress
+	} else {
+		target = *tx.To()
+	}
+
+	// Signature type
+	data.WriteByte(byte(sigtype)) // 1 byte: 00 == EOACreate, 01 == EIP 155, 02 == ETH Sign Message
```
Comment (tynes) [inline core/state_transition_ovm.go]: I've played with a few different styles of serialization, using `bytes.Buffer`, `io.Writer` and `[]byte`. Curious if you have any preferences.

My favorite so far has been using `[]byte` directly, see https://github.com/ethereum-optimism/go-ethereum/blob/b2e6790696a97e5e7c7c0c7f6facd4b66478d51f/rollup/types.go#L382

File: core/state_transition_ovm.go (hunk 10 above)
Comment (tynes) [reply core/state_transition_ovm.go:196]: I think we updated them, see https://github.com/ethereum-optimism/go-ethereum/blob/b2e6790696a97e5e7c7c0c7f6facd4b66478d51f/core/types/transaction.go#L42-L45

File: core/state_transition_ovm.go [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+	if msg.SignatureHashType() == 0 {
+		return 1
+	} else if msg.SignatureHashType() == 1 {
+		return 2
+	} else {
+		return 0
+	}
+}
+
+func getQueueOrigin(
```
Comment (tynes) [inline core/state_transition_ovm.go:208]: I'm sorry about this

File: core/vm/evm.go (hunk 12 above)
Comment (tynes) [reply core/vm/evm.go]: This would be a good comment to leave directly in the code

File: core/vm/evm.go [hunk 21]
```diff
@@ -69,6 +88,7 @@ func run(evm *EVM, contract *Contract, input []byte, readOnly bool) ([]byte, err
 			return RunPrecompiledContract(p, input, contract)
 		}
 	}
+
```
Comment (tynes) [inline core/vm/evm.go:98]: Maybe no need to add extra whitelines for minimal diff?

File: core/vm/evm.go (hunk 15 above)
Comment (tynes) [reply core/vm/evm.go]: Is there any chance `len(ret) < 96`?

File: core/vm/ovm_state_dump.go [hunk 22]
```diff
@@ -0,0 +1,3623 @@
...
+	var err error
+
+	err = json.Unmarshal(ovmStateDumpJSON, &OvmStateDump)
+	if err != nil {
+		panic(fmt.Errorf("could not decode OVM state dump: %v", err))
+	}
+
+	OvmStateManager = OvmStateDump.Accounts["OVM_StateManager"]
+	OvmExecutionManager = OvmStateDump.Accounts["OVM_ExecutionManager"]
+	UsingOVM = os.Getenv("USING_OVM") == "true"
```
Comment (tynes) [inline core/vm/ovm_state_dump.go:3686]: Noted that this will be added to deploy config

File: internal/ethapi/api.go (hunk 17 above)
Comment (tynes) [reply internal/ethapi/api.go:809]: I'm using `ExecutionManager.GetMaxTransactionGasLimit`

File: core/state_transition.go (hunk 4 above)
Comment (karlfloersch) [reply core/state_transition.go:284]: Once we've stabilized Synthetix running in Contracts v2, and have deployed Mintr to this Geth & have it work, that seems to me to be the next highest priority

File: core/types/transaction.go (hunk 11 above)
Comment (karlfloersch) [reply core/types/transaction.go:335]: I'm not entirely sure but I think it looks about right tbh.

Also, right here you should add:

```
if tx.meta.lastL1BlockNumber != nil {
  msg.l1BlockNumber = tx.meta.lastL1BlockNumber
}
```

Oh though I forgot -- we have to rename `txId` here - https://github.com/ethereum-optimism/go-ethereum/pull/67/files#diff-5500355cc4409489bee920d34f30a486e081409325cc12793ea7895fb64ac7daR605 to `lastL1BlockNumber`

File: core/state_transition_ovm.go (hunk 7 above)
Comment (karlfloersch) [reply core/state_transition_ovm.go:46]: > For an L1 to L2 transaction, is block.number supposed to be the L1 block number?

Yep

> @karlfloersch @tynes what's the correct value for this?

Ah so we've got to set it in the same way you set the L1MessageSender from the meta. I wrote up how that works below

**Review by karlfloersch — COMMENTED** [review]
Got part of the way through & will finish shortly but so far it's looking great!

File: core/state_transition_ovm.go [hunk 23]
```diff
@@ -0,0 +1,190 @@
...
+	if err != nil {
+		return msg, err
+	}
+
+	if msg.From() == GodAddress {
+		return msg, nil
+	}
+
+	v, r, s := tx.RawSignatureValues()
+	v = big.NewInt(int64(v.Uint64() - 35 - 2*420))
```
Comment (gakonst) [inline core/state_transition_ovm.go]:
```suggestion
	v = big.NewInt(int64(v.Uint64() - 35 - 2* CHAIN_ID))
```

Could refactor this out to a CHAIN_ID constant, maybe also add a comment that this is EIP-155?

File: core/state_transition.go [hunk 24]
```diff
@@ -149,6 +138,18 @@ func NewStateTransition(evm *vm.EVM, msg Message, gp *GasPool) *StateTransition
...
+			return nil, 0, false, err
+		}
+
+		evm.Context.Origin = msg.From()
+
+		return NewStateTransition(evm, ovmmsg, gp).TransitionDb()
+	}
+
 	return NewStateTransition(evm, msg, gp).TransitionDb()
 }
```
Comment (gakonst) [inline core/state_transition.go]: Believe you could do this instead?

```suggestion
	if vm.UsingOVM {
		// OVM_ENABLED
		msg, err := toExecutionManagerRun(evm, msg)
		if err != nil {
			return nil, 0, false, err
		}
		evm.Context.Origin = msg.From()
	}	

	return NewStateTransition(evm, msg, gp).TransitionDb()
}
```

File: core/state_transition.go (hunk 2 above)
Comment (gakonst) [reply core/state_transition.go]: Why is it safe to skip the nonce check?

File: core/state_transition.go (hunk 3 above)
Comment (gakonst) [reply core/state_transition.go:252]: awesome refactor

File: core/vm/evm.go [hunk 25]
```diff
@@ -47,14 +47,40 @@ type (
...
+				isUnknown = false
+				abi := &(account.ABI)
+				method, err := abi.MethodById(input)
+				if err != nil {
+					log.Debug("Calling Known Contract", "Name", name, "ERROR", err)
+				} else {
+					log.Debug("Calling Known Contract", "Name", name, "Method", method.RawName)
+				}
+			}
+		}
```
Comment (gakonst) [inline core/vm/evm.go:67]: `OvmStateDump.Accounts` is already a [map](https://github.com/ethereum-optimism/go-ethereum/blob/dd0caa08486314f00c2422231e134ffc77d45d21/core/vm/ovm_state_dump.go#L3663), can't you just do `account, ok := OvmStateDumpAccounts[name]`?

File: core/vm/evm.go [hunk 26]
```diff
@@ -47,14 +47,40 @@ type (
...
+
+		// Temporary: Safety checker always returns true.
+		// ovmTODO: Remove this.
+		if contract.Address() == OvmStateDump.Accounts["OVM_SafetyChecker"].Address {
+			return AbiBytesTrue, nil
+		}
+
+		// If we're calling the state manager, we want to use our native implementation instead.
+		if contract.Address() == OvmStateManager.Address {
+			return callStateManager(input, evm, contract)
```
Comment (gakonst) [inline core/vm/evm.go:82]: Just to be clear, will no user ever _directly_ call the OvmStateManager? Will it always be an internal call made by the `ExecutionManager.run`'s `ovmCALL` which will execute the ovmSolc re-written `ovmSSTORE` etc.?

File: core/vm/evm.go (hunk 13 above)
Comment (gakonst) [reply core/vm/evm.go:250]: Why? What do these mean? Could we use a variable instead of hardcoded addrs?

File: core/vm/ovm_state_manager.go [hunk 27]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+func owner(evm *EVM, contract *Contract, args map[string]interface{}) ([]interface{}, error) {
+	origin := evm.Context.Origin
+
+	return []interface{}{origin}, nil
+}
+
+func setAccountNonce(evm *EVM, contract *Contract, args map[string]interface{}) ([]interface{}, error) {
+	address := args["_address"].(common.Address)
```
Comment (gakonst) [inline core/vm/ovm_state_manager.go]:
```suggestion
	address, ok := args["_address"].(common.Address)
```

I wonder if a client may be able to provide "bad arguments" in a way and cause a panic here? May be better to handle the error (same for all other interface -> type casts)

File: internal/ethapi/api.go [hunk 28]
```diff
@@ -830,7 +830,11 @@ func DoCall(ctx context.Context, b Backend, args CallArgs, blockNrOrHash rpc.Blo
 	}
 
 	// Create new call message
-	msg := types.NewMessage(addr, args.To, 0, value, gas, gasPrice, data, false, nil, nil, types.QueueOriginSequencer, 0)
+	msg := types.NewMessage(addr, args.To, 0, value, gas, gasPrice, data, false, &addr, nil, types.QueueOriginSequencer, 0)
+	outmsg, err := core.EncodeFakeMessage(msg)
```
Comment (gakonst) [inline internal/ethapi/api.go]: Why is this required?

File: core/state_transition_ovm.go [hunk 29]
```diff
@@ -0,0 +1,190 @@
...
+	msg Message,
+) (Message, error) {
+	var input = []interface{}{
+		big.NewInt(int64(msg.Gas())),
+		msg.To(),
+		msg.Data(),
+	}
+
+	var abi = vm.OvmStateDump.Accounts["mockOVM_ECDSAContractAccount"].ABI
+	output, err := abi.Pack("qall", input...)
```
Comment (gakonst) [inline core/state_transition_ovm.go:151]: This is from [here](https://github.com/ethereum-optimism/contracts-v2/pull/33/files#diff-cb33f9d006541cbc42ba5d32c56da5a9820d7649e8e286072195c75f5422dc3fR83)? Is there some explanation on how this kind of logic works?

File: core/vm/evm.go [hunk 30]
```diff
@@ -238,12 +299,40 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
...
+		// OVM_ENABLED
+		if evm.EthCallSender != nil && *evm.EthCallSender == addr {
+			// We have to handle eth_call in a special manner as it doesn't stem from a signed
+			// transaction and therefore can't go through the standard EOA contract. When we detect
+			// this case, we temporarily insert some mock code that allows the user to pass through
+			// the EOA without a signature. EthCallSender should *never* be made non-nil outside
+			// of an eth_call. We store the old code here so that we're able to reset the code to
+			// the original code for the remainder of the transaction.
+			prevCode = evm.StateDB.GetCode(addr)
+			evm.StateDB.SetCode(addr, common.FromHex(OvmStateDump.Accounts["mockOVM_ECDSAContractAccount"].Code))
```
Comment (gakonst) [inline core/vm/evm.go:319]: What do we expect the code at `mockOVM_ECDSAContractAccount` to be?

File: core/vm/evm.go [hunk 31]
```diff
@@ -238,12 +299,40 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
...
+		}
+	}
+
 	// Initialise a new contract and set the code that is to be used by the EVM.
 	// The contract is a scoped environment for this execution context only.
 	contract := NewContract(caller, to, value, gas)
 	contract.SetCallCode(&addr, evm.StateDB.GetCodeHash(addr), evm.StateDB.GetCode(addr))
```
Comment (gakonst) [inline core/vm/evm.go:326]: Since the value is reset here below, why is L319 needed?

Could we instead do the following (pseudocode) instead of doing the temp-value trick above?

```
callcode := common.FromHex(OvmStateDump.Accounts["mockOVM_ECDSAContractAccount"].Code)
codehash := hash_fn(callcode)
contract.SetCallCode(&addr, codehash, callcode)
```

That would also remove the need of doing L332 too?

File: core/vm/evm.go [hunk 32]
```diff
@@ -471,14 +627,26 @@ func (evm *EVM) create(caller ContractRef, codeAndHash *codeAndHash, gas uint64,
...
+			log.Error("Creation called by non-Execution Manager contract! This should never happen.", "Offending address", caller.Address().Hex())
+			return nil, caller.Address(), 0, ErrOvmCreationFailed
+		}
+
+		// ovmADDRESS will be set by the execution manager to the target address whenever it's
+		// about to create a new contract. This value is currently stored at the [15] storage slot.
+		// Can pull this specific storage slot to get the address that the execution manager is
+		// trying to create to, and create to it.
+		slot := common.HexToHash(strconv.FormatInt(15, 16))
+		contractAddr = common.BytesToAddress(evm.StateDB.GetState(OvmExecutionManager.Address, slot).Bytes())
```
Comment (gakonst) [inline core/vm/evm.go:645]: Is this referring to [this](https://github.com/ethereum-optimism/contracts-v2/blob/master/contracts/optimistic-ethereum/iOVM/execution/iOVM_ExecutionManager.sol#L63)? This is maybe a footgun in case the layout of the ExecutionManager changes, maybe consider adding a big big warning to this _and_ the contracts?

File: core/vm/interpreter.go (hunk 16 above)
Comment (gakonst) [reply core/vm/interpreter.go:127]: Curious, what was the bug - also why do we need to do this manually? Is`in.returnData` not always set correctly?

File: internal/ethapi/api.go [hunk 33]
```diff
@@ -859,7 +863,10 @@ func DoCall(ctx context.Context, b Backend, args CallArgs, blockNrOrHash rpc.Blo
 	// Setup the gas pool (also for unmetered requests)
 	// and apply the message.
 	gp := new(core.GasPool).AddGas(math.MaxUint64)
-	res, gas, failed, err := core.ApplyMessage(evm, msg, gp)
+	log.Debug(">>>>>> Serving an eth_call <<<<<<", "From", addr.Hex(), "To", args.To.Hex())
```
Comment (gakonst) [inline internal/ethapi/api.go]:
```suggestion
```

File: internal/ethapi/api.go [hunk 34]
```diff
@@ -859,7 +863,10 @@ func DoCall(ctx context.Context, b Backend, args CallArgs, blockNrOrHash rpc.Blo
...
+	evm.Context.EthCallSender = &addr
+	res, gas, failed, err := core.ApplyMessage(evm, outmsg, gp)
+	log.Debug("<<<<<< Served an eth_call  >>>>>>", "From", addr.Hex(), "To", args.To.Hex())
```
Comment (gakonst) [inline internal/ethapi/api.go]:
```suggestion
```

(just a reminder to remove once done debugging)

File: core/vm/evm.go (hunk 32 above)
Comment (gakonst) [reply core/vm/evm.go:645]: Also, consider refactoring the `else` branch to a helper function?

File: core/vm/evm.go [hunk 35]
```diff
@@ -266,6 +356,65 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
 			contract.UseGas(contract.Gas)
 		}
 	}
+
+	if UsingOVM {
```
Comment (gakonst) [inline core/vm/evm.go:360]: What is the rationale behind the whole isTarget/OriginalTarget logic? What can go wrong here?

File: core/vm/interpreter.go (hunk 16 above)
Comment (karlfloersch) [reply core/vm/interpreter.go:127]: This is a pretty crazy bug 😮

Has to do with internal Geth memory management... we gotta look into it further

File: core/state_transition.go (hunk 2 above)
Comment (smartcontracts) [reply core/state_transition.go]: It wasn't :-( I added the nonce check again.

File: core/vm/evm.go (hunk 13 above)
Comment (smartcontracts) [reply core/vm/evm.go:250]: `0xdead` addresses are used for all of the OVM contracts that run in L2 (so the execution manager, state manager, etc.) Users aren't allowed to call these addresses via ovmCALLs & similar opcodes. We could probably use a variable but we need to coordinate this variable across everything.

File: core/vm/evm.go (hunk 15 above)
Comment (smartcontracts) [reply core/vm/evm.go]: Yes -- fixed to address this.

File: core/vm/interpreter.go (hunk 16 above)
Comment (smartcontracts) [reply core/vm/interpreter.go:127]: @gakonst unclear what the bug is. It appears that some combination of `RETURNDATACOPY` and other opcodes results in `MSTORE` overwriting `in.returnData`. Very strange.

File: core/state_transition_ovm.go (hunk 18 above)
Comment (smartcontracts) [reply core/state_transition_ovm.go:19]: Took this from another part of geth where it used the escaped `"`.

File: core/state_processor.go [hunk 36]
```diff
@@ -82,20 +83,43 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
...
+		// OVM_ENABLED
+		log.Debug(">>>>>> Serving an OVM transaction <<<<<<")
+	}
+
 	// Apply the transaction to the current state (included in the env)
 	_, gas, failed, err := ApplyMessage(vmenv, msg, gp)
+
+	if vm.UsingOVM {
+		// OVM_ENABLED
+		log.Debug("<<<<<< Served an OVM transaction  >>>>>>")
```
Comment (tynes) [inline core/state_processor.go]: I don't know if we need these loglines anymore, but it probably isn't a bad idea to log something along the lines of `ovm enabled` once at startup if the env var that turns on ovm is set

File: core/state_processor.go [hunk 37]
```diff
@@ -82,20 +83,43 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
...
+	}
+
 	if err != nil {
 		return nil, err
 	}
+
```
Comment (tynes) [inline core/state_processor.go]: I don't think we need to add extra newlines as it just creates a bigger diff from upstream. Go code generally has less newlines between blocks of code compared to JS based on reading a bunch of Go codebases 🤷

File: core/state_transition_ovm.go [hunk 38]
```diff
@@ -0,0 +1,229 @@
...
+	} else if msg.SignatureHashType() == 1 {
+		return 2
+	} else {
+		return 1
+	}
+}
+
+func getQueueOrigin(
+	queueOrigin *big.Int,
+) (types.QueueOrigin, error) {
```
Comment (tynes) [inline core/state_transition_ovm.go:210]: I'm confused about this because there should only be 2 queue origins?

File: core/vm/ovm_state_manager.go (hunk 27 above)
Comment (tynes) [reply core/vm/ovm_state_manager.go]: I agree and think its a better idea to be thorough with the error checking

File: core/vm/evm.go (hunk 26 above)
Comment (karlfloersch) [reply core/vm/evm.go:82]: Yep no one can directly call the state manager. This is because it's a `dead` address & therefore not able to be called directly

**Comment by tynes** [comment]
https://github.com/ethereum-optimism/optimism-monorepo/pull/358 must be merged at the same time

File: internal/ethapi/api.go (hunk 33 above)
Comment (tynes) [reply internal/ethapi/api.go]: Definitely should remove this logline, there will be tons of these

File: core/state_transition_ovm.go [hunk 39]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+func init() {
+	// ovmTODO: Pass this in via standard config flow instead of via environment variables.
+	// kelvin's note: "tee hee, sorry!"
+	address := os.Getenv("TX_INGESTION_SIGNER_ADDRESS")
+
+	if len(address) == 0 {
+		log.Warn("No TX_INGESTION_SIGNER_ADDRESS supplied. Using ZERO_ADDRESS default.")
+		address = "0000000000000000000000000000000000000000"
```
Comment (tynes) [inline core/state_transition_ovm.go:35]: This always breaks in this code path because address len is 40
